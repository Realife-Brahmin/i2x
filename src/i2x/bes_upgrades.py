# Copyright (C) 2023 Battelle Memorial Institute
# file: bes_upgrades.py
# from a saved HCA solution, estimates the limiting branch upgrades

import sys
import numpy as np
import i2x.mpow_utilities as mpow
import json
import math

def get_default_line_mva (kv):
  if kv <= 121.0:
    return 131.0
  elif kv <= 145.0:
    return 157.0
  elif kv <= 242.0:
    return 600.0
  elif kv <= 362.0:
    return 1084.0
  elif kv <= 550.0:
    return 1800.0
  return -1.0 # force an error

def get_default_transformer_mva (kv):
  if kv <= 121.0:
    return 100.0
  elif kv <= 145.0:
    return 200.0
  elif kv <= 242.0:
    return 400.0
  elif kv <= 362.0:
    return 800.0
  elif kv <= 550.0:
    return 1200.0
  return -1.0 # force an error

def reset_mva (d, i, mva):
  newval = '{:.2f}'.format(mva)
  d['branch'][i][mpow.RATE_A] = newval
  d['branch'][i][mpow.RATE_B] = newval
  d['branch'][i][mpow.RATE_C] = newval

def estimate_overhead_line_length (x, kv):
  if kv > 242.0:
    return x / 0.6
  return x / 0.8

def estimate_overhead_lines_in_parallel (z, kv):
  zsurge = 400.0
  if kv > 242.0:
    zsurge = 300.0
  npar = int (0.5 + zsurge / z)
  if npar < 1:
    npar = 1
  return npar

def estimate_transformers_in_parallel (mva, kv):
  npar = int (0.5 + mva / get_default_transformer_mva(kv))
  if npar < 1:
    npar = 1
  return npar

def get_parallel_branch_scale (npar):
  if npar > 1:
    return 1.0 / float(npar)
  return 0.0

def get_branch_description (branch, bus, i):
  desc = '??'
  bus1 = int(branch[i,mpow.F_BUS])
  bus2 = int(branch[i,mpow.T_BUS])
  kv1 = bus[bus1-1,mpow.BASE_KV]
  kv2 = bus[bus2-1,mpow.BASE_KV]
  if kv1 > 100.0 and kv2 > 100.0:
    xpu = branch[i,mpow.BR_X]
    if branch[i,mpow.TAP] > 0.0:
      mva = 100.0 * 0.10 / xpu
      desc = 'Xfmr {:3d}-{:3d} {:7.2f} / {:7.2f} kV x={:.4f}, mva={:.2f}'.format (bus1, bus2, kv1, kv2, xpu, mva)
    elif xpu > 0.0:
      bpu = branch[i,mpow.BR_B]
      npar = 1
      if bpu > 0.0:
        zbase = kv1*kv1/100.0
        x = xpu*zbase
        xc = zbase / branch[i,mpow.BR_B]
        z = math.sqrt(x * xc)
        npar = estimate_overhead_lines_in_parallel (z, kv1)
      mva = npar * get_default_line_mva (kv1)
      # TODO: need a test for overhead vs cable
      miles = estimate_overhead_line_length (x*npar, kv1)
      desc = 'Line {:3d}-{:3d} {:7.2f} kV x={:.4f}, z={:6.2f} ohms, npar={:d}, mva={:.2f}, mi={:.2f}'.format (bus1, bus2, kv1, xpu, z, npar, mva, miles)
    elif xpu < 0.0:
      mva = get_default_line_mva (kv1)
      desc = 'Scap {:3d}-{:3d} {:7.2f} kV x={:.4f}, mva={:.2f}'.format (bus1, bus2, kv1, xpu, mva)
  return desc

def set_estimated_branch_ratings (matpower_dictionary, branch_contingencies=None, contingency_mva_threshold=100.0):
  bus = np.array (matpower_dictionary['bus'], dtype=float)
  branch = np.array (matpower_dictionary['branch'], dtype=float)
  nb = len(bus)
  nl = len(branch)

  for i in range(nl):
    bus1 = int(branch[i,mpow.F_BUS])
    bus2 = int(branch[i,mpow.T_BUS])
    kv1 = bus[bus1-1,mpow.BASE_KV]
    kv2 = bus[bus2-1,mpow.BASE_KV]
    if kv1 > 100.0 and kv2 > 100.0:
      scale = 0.0
      # mva = branch[i,mpow.RATE_A] these are invalid for IEEE118 and WECC240
      xpu = branch[i,mpow.BR_X]
      if branch[i,mpow.TAP] > 0.0:
        mva = 100.0 * 0.10 / xpu
        npar = estimate_transformers_in_parallel (mva, max(kv1, kv2))
        scale = get_parallel_branch_scale (npar)
        reset_mva (matpower_dictionary, i, mva)
      elif xpu > 0.0:
        bpu = branch[i,mpow.BR_B]
        npar = 1
        if bpu > 0.0:
          zbase = kv1*kv1/100.0
          x = xpu*zbase
          xc = zbase / branch[i,mpow.BR_B]
          z = math.sqrt(x * xc)
          npar = estimate_overhead_lines_in_parallel (z, kv1)
          scale = get_parallel_branch_scale (npar)
        mva = npar * get_default_line_mva (kv1)
        reset_mva (matpower_dictionary, i, mva)
      elif xpu < 0.0: # not considering series capacitors in parallel, which is valid for the WECC 240-bus model
        mva = get_default_line_mva (kv1)
        reset_mva (matpower_dictionary, i, mva)
      if branch_contingencies is not None:
        if mva >= contingency_mva_threshold:
          branch_contingencies.append({'branch':i+1, 'scale':scale})
#          print ('contingency {:s}'.format (get_branch_description (branch, bus, i)))

#TODO: need functions that:
#  1) return a dictionary of the limiting branches and their parameters, def get_limiting_branches (case_file, results_file):
#  2) return branch parameters by index, def get_branch_attributes (case_file, idx):
def print_limiting_branches (case_file, results_file):
  d = mpow.read_matpower_casefile (case_file)
  branch = np.array (d['branch'], dtype=float)
  bus = np.array (d['bus'], dtype=float)

  lp = open (results_file).read()
  r = json.loads(lp)
  print (' Bus   HC[GW]')
  for key, val in r['buses'].items():
    print ('{:4d} {:8.3f}'.format (int(key), val['fuels']['hca']))
    i_max = val['max_max_muF']['branch']
    mu_max = val['max_max_muF']['muF']
    i_mean = val['max_mean_muF']['branch']
    mu_mean = val['max_mean_muF']['muF']
    print ('       Max Mu Branch: {:4d} ({:8.3f}) {:s}'.format (i_max, mu_max, 
                                                                get_branch_description (branch, bus, i_max)))
    print ('      Mean Mu Branch: {:4d} ({:8.3f}) {:s}'.format (i_mean, mu_mean, 
                                                                get_branch_description (branch, bus, i_mean)))

