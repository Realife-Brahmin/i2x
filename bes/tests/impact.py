# Copyright (C) 2023 Battelle Memorial Institute
# file: impact.py
"""Simulate a sequence of system impact studies.
"""
import os
import numpy as np
import i2x.bes_hca as hca
import i2x.bes_upgrades as bes
import i2x.mpow_utilities as mpow

bes_size = 5000.0
bes_buses = [6, 3, 8]
HCA_PMAX = 30000.0
HCA_QMAX = 10000.0

bWind = False # alternate addition of wind and solar projects

def report_branch_limits (r, bus, branch, bLog=False):
  brnum = 0
  scale = 0.0
  cost = 0.0
  # merit order of branch upgrades based on shadow price
  d = {}
  if bLog:
    print ('Branches at limit:')
  for key, val in r['branches'].items():
    muF = val['muF']
    d[key] = muF
    if bLog:
      print ('  {:4d} {:s} {:s}'.format (key, str(val), bes.get_branch_description (branch, bus, key-1, bEstimateCost=True)))
  if len(d) > 0:
    merit = sorted (d.items(), key=lambda x:x[1], reverse=True)
    brnum = merit[0][0]
    scale, cost, miles = bes.get_branch_next_upgrade (branch, bus, brnum-1)
    if bLog:
      print ('Merit Order of Upgrades:', merit)
      print ('Upgrading', brnum, bes.get_branch_description (branch, bus, brnum-1, bEstimateCost=True))
  return brnum, scale, cost

def load_starting_case ():
  # read the exisitng network with valid MVA ratings and generator costs
  mpd = mpow.read_matpower_casefile ('hca_wmva.m', asNumpy=True)

  # add the candidate generator for HCA
  mpd['gen'] = np.vstack ((mpd['gen'], 
                          np.array([1.0,0.0,0.0,HCA_QMAX,-HCA_QMAX,1.0,1000.0,0.0,HCA_PMAX,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0])))
  mpd['gentype'].append('DL')
  mpd['genfuel'].append('hca')
  mpd['gencost'] = np.vstack((mpd['gencost'], mpow.get_hca_gencosts('hca')))

  # extra generator data (xgd) including the new one for HCA
  # assume all units have been on for 24 hours to start, so MOST can leave them on or switch them off without restriction
  unit_states = np.ones(len(mpd['gen'])) * 24.0
  mpow.write_xgd_function ('hca_xgd', mpd['gen'], mpd['gencost'], mpd['genfuel'], unit_states, bLog=False)

  G = bes.build_matpower_graph (mpd)
  mpow.write_matpower_casefile (mpd, 'hca_case')
  return mpd, G

def add_renewable (poc, bes_size, ic_cost):
  global bWind
  if bWind:
    genfuel = 'wind'
    gentype = 'WT'
  else:
    genfuel = 'solar'
    gentype = 'PV'
    bWind = True
  qmax = 0.25 * bes_size
  sbase = min(1000.0, bes_size)
  pg = 0.5 * bes_size
  qg = 0.0
  pmin = 0.0
  mpd['gen'] = np.vstack ((mpd['gen'], 
                          np.array([poc,pg,qg,qmax,-qmax,1.0,sbase,0.0,bes_size,pmin,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0])))
  mpd['gentype'].append(gentype)
  mpd['genfuel'].append(genfuel)
  mpd['gencost'] = np.vstack((mpd['gencost'], mpow.get_hca_gencosts(genfuel)))

  unit_states = np.ones(len(mpd['gen'])) * 24.0
  mpow.write_xgd_function ('hca_xgd', mpd['gen'], mpd['gencost'], mpd['genfuel'], unit_states, bLog=False)

  mpow.write_matpower_casefile (mpd, 'hca_case')
  print ('  Added {:.2f} MW {:s} at bus {:d} for ${:.3f}M'.format (bes_size, genfuel, poc, cost/1.0e6))
  return mpd

def update_current_case (mpd, brnum, scale):
  idx = brnum-1
  br = mpd['branch']
  br[idx][mpow.BR_R] /= scale
  br[idx][mpow.BR_X] /= scale
  br[idx][mpow.BR_B] *= scale
  br[idx][mpow.RATE_A] *= scale
  br[idx][mpow.RATE_B] *= scale
  br[idx][mpow.RATE_C] *= scale
  G = bes.build_matpower_graph (mpd)
  mpow.write_matpower_casefile (mpd, 'hca_case')
  return mpd, G

def process_renewable_project (poc, mw_size, mpd, G, load_scale=2.75, itmax=3):
  print ('Processing application for {:.2f} MW at bus {:d}'.format (mw_size, poc))
  itcount = 0
  mw_hc = 0.0
  total_cost = 0.0
  while mw_hc < mw_size and itcount < itmax:
    itcount += 1
    cont = bes.rebuild_contingencies (mpd, G, [poc])
    mpow.write_contab_list ('hca_contab', mpd, cont, bLog=False)
    r = hca.bes_hca_fn (sys_name='hca', case_title='hca', load_scale=load_scale, hca_buses=[poc])
    mw_hc = r['buses'][poc]['fuels']['hca']*1000.0
    if mw_hc < mw_size:
      brnum, scale, cost = report_branch_limits (r, mpd['bus'], mpd['branch'])
      if brnum > 0:
        print ('  HC={:.2f} MW, upgrade branch {:d} by {:.4f} at ${:.3f}M'.format (mw_hc, brnum, scale, cost/1.0e6))
        mpd, G = update_current_case (mpd, brnum, scale)
        total_cost += cost

  return mpd, G, total_cost, mw_hc

def rollback_grid_upgrades (mpd, pre_project_branches):
  mpd['branch'] = np.copy (pre_project_branches)
  G = bes.build_matpower_graph (mpd)
  mpow.write_matpower_casefile (mpd, 'hca_case')
  return mpd, G

def process_auction (poc_buses, load_scale=2.75):
  print ('***********************************************')
  print ('Auction Process on buses', str(poc_buses))
  mpd = mpow.read_matpower_casefile ('hca_wmva.m', asNumpy=True)

  # add the candidate generators at each poc bus
  for bus in poc_buses:
    mpd['gen'] = np.vstack ((mpd['gen'], 
                             np.array([bus,0.0,0.0,HCA_QMAX,-HCA_QMAX,1.0,1000.0,0.0,HCA_PMAX,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0])))
    mpd['gentype'].append('DL')
    mpd['genfuel'].append('hca')
    mpd['gencost'] = np.vstack((mpd['gencost'], mpow.get_hca_gencosts('hca')))

  # extra generator data (xgd) including the new one for HCA
  # assume all units have been on for 24 hours to start, so MOST can leave them on or switch them off without restriction
  unit_states = np.ones(len(mpd['gen'])) * 24.0
  mpow.write_xgd_function ('hca_xgd', mpd['gen'], mpd['gencost'], mpd['genfuel'], unit_states, bLog=False)

  G = bes.build_matpower_graph (mpd)
  mpow.write_matpower_casefile (mpd, 'hca_case')

  cont = bes.rebuild_contingencies (mpd, G, poc_buses)
  mpow.write_contab_list ('hca_contab', mpd, cont, bLog=False)

  print ('Generation by Fuel[GW]')
  print (' '.join(['{:>7s}'.format(x) for x in mpow.FUEL_LIST]))
  fscript, fsummary = mpow.write_hca_solve_file ('hca', load_scale=load_scale, quiet=False)

  if os.path.exists(fsummary):
    os.remove(fsummary)

  mpow.run_matpower_and_wait (fscript, quiet=True)

  if not os.path.exists(fsummary):
    print ('** Auction HCA solution failed **')
    return

  f, nb, ng, nl, ns, nt, nj_max, nc_max, psi, Pg, Pd, Rup, Rdn, SoC, Pf, u, lamP, muF = mpow.read_most_solution(fsummary)
  meanPg = np.mean(Pg[:,0,0,:], axis=1)
  meanPd = np.mean(Pd[:,0,0,:], axis=1)
  meanPf = np.mean(Pf[:,0,0,:], axis=1)
  meanlamP = np.mean(lamP[:,0,0,:], axis=1)
  meanmuF = np.mean(muF[:,0,0,:], axis=1)
  baselamP = lamP[:,0,0,0]
  basemuF = muF[:,0,0,0]
  actualPd = np.sum(np.mean(Pd[:,0,0,:], axis=1))
  meanPgen = np.mean(Pg[:,0,0,:], axis=1)
  actualPgen = np.sum(meanPgen)

  # summarize the generation by fuel type
  fuel_Pg = {}
  for fuel in mpow.FUEL_LIST:
    fuel_Pg[fuel] = 0.0
  for i in range(ng):
    fuel = mpd['genfuel'][i]
    fuel_Pg[fuel] += meanPgen[i]
  for fuel in mpow.FUEL_LIST:
    fuel_Pg[fuel] *= 0.001
  fuel_str = ' '.join(['{:7.3f}'.format(fuel_Pg[x]) for x in mpow.FUEL_LIST])
  print (fuel_str)

  print (' Bus    HC [MW]')
  for i in range (len(mpd['genfuel'])):
    if mpd['genfuel'][i] == 'hca':
      mw = meanPg[i]
      bus = mpd['gen'][i][mpow.GEN_BUS]
      print ('{:4d} {:10.3f}'.format (int(bus), mw))

  print ('Merit order of upgrades:')
  print (' Br# From   To     muF      MVA      kV  Add MVA    Miles  Cost $M')
  r = {}
  for i in range(nl):
    if meanmuF[i] > 0.0:
      r[i+1] = meanmuF[i]
  merit = sorted (r.items(), key=lambda x:x[1], reverse=True)
  for pair in merit:
    i = pair[0]
    muF = pair[1]
    rating = mpd['branch'][i-1][mpow.RATE_A]
    fbus = int(mpd['branch'][i-1][mpow.F_BUS])
    tbus = int(mpd['branch'][i-1][mpow.T_BUS])
    kv = max (mpd['bus'][fbus-1][mpow.BASE_KV], mpd['bus'][tbus-1][mpow.BASE_KV])
    scale, cost, miles = bes.get_branch_next_upgrade (mpd['branch'], mpd['bus'], i-1)
    if miles < 0.0:
      label = 'Scap'
    elif miles > 0.0:
      label = '{:8.2f}'.format (miles)
    else:
      label = 'Xfmr'
    addmva = rating * (scale - 1.0)
    print ('{:4d} {:4d} {:4d} {:7.4f} {:8.2f} {:7.2f} {:8.2f} {:8s} {:8.2f}'.format(i, fbus, tbus, muF, rating, kv, addmva, label, cost/1.0e6))

if __name__ == '__main__':
  process_auction ([6, 3, 8])

  queue = [{'poc':6, 'mw': 5000.0, 'itlim': 3, 'costlim': 1000.0e6},
           {'poc':3, 'mw': 4400.0, 'itlim': 3, 'costlim':  500.0e6},
           {'poc':8, 'mw': 2000.0, 'itlim': 3, 'costlim': 1300.0e6}
          ]
  print ('***********************************************')
  print ('Queue Process on {:d} applications'.format (len(queue)))
  mpd, G = load_starting_case()

  for app in queue:
    pre_project_branches = np.copy (mpd['branch']) # in case we have to roll back upgrades on projects that don't move forward
    mpd, G, cost, hc = process_renewable_project (app['poc'], app['mw'], mpd, G)
    if hc >= app['mw']:
      if cost <= app['costlim']:
        mpd = add_renewable (app['poc'], app['mw'], cost)
      else:
        print ('  IX Cost ${:.3f}M is too high, HC={:.3f} MW'.format (cost/1.0e6, hc))
        if cost > 0.0:
          mpd, G = rollback_grid_upgrades (mpd, pre_project_branches)
    else:
      print ('  HC {:.3f} MW is too low'.format (hc))
      if cost > 0.0:
        mpd, G = rollback_grid_upgrades (mpd, pre_project_branches)

