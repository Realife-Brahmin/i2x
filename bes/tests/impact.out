total_time = 1.0378
***********************************************
Auction Process on hca_wmva.m, POC buses=[6, 3, 8]
Generation by Fuel[GW]
    hca    wind   solar nuclear   hydro    coal      ng      dl
  9.648  14.391   0.000   5.139   0.000  21.566  15.341   0.000
 Bus    HC [MW]
   6   4552.694
   3   3855.826
   8   1239.512
Merit order of upgrades:
 Br# From   To     muF      MVA      kV  Add MVA    Miles  Cost $M
   1    5    6  4.8935  2168.00  345.00  1084.00   142.43   508.49
  13    1    3  4.1832  3252.00  345.00  1084.00   212.56   753.96
   4    1    2  4.0474  2168.00  345.00  1084.00   207.00   734.49
   8    6    7  3.5878  2168.00  345.00  1084.00   199.86   709.52
   7    4    8  1.7354  2168.00  345.00  1084.00   214.74   761.59
   9    2    5  0.3557  6504.00  345.00  1084.00   148.51   529.78
  11    3    4  0.0190  2168.00  345.00  1084.00   147.63   526.71
***********************************************
Queue Process on 3 applications for hca_wmva.m
Processing application for 5000.00 MW at bus 6
  Added 5000.00 MW solar at bus 6 for $0.000M
Processing application for 4400.00 MW at bus 3
  HC=4122.50 MW, upgrade branch 1 by 1.5000 at $508.489M
  IX Cost $508.489M is too high, HC=4422.132 MW
Processing application for 2000.00 MW at bus 8
  HC=1239.51 MW, upgrade branch 1 by 1.5000 at $508.489M
  HC=1239.51 MW, upgrade branch 7 by 1.5000 at $761.593M
  Added 2000.00 MW wind at bus 8 for $1270.082M
